perspective.el

```diff
@@ -719,10 +720,43 @@ named collections of buffers and window configurations."
     (ad-deactivate-regexp "^persp-.*")
     (remove-hook 'after-make-frame-functions 'persp-init-frame)
     (remove-hook 'ido-make-buffer-list-hook 'persp-set-ido-buffers)
+    (delete 'persp-frame-deleted 'delete-frame-functions)
     (setq read-buffer-function nil)
     (setq perspectives-hash nil)
     (setq global-mode-string (delq 'persp-modestring global-mode-string))))
 
+(defvar persp-saved-hash nil
+  "Stores perspective from the last deleted frame.")
+
+(defun persp-frame-deleted (frame)
+  "Import or save perspectives from a deleted frame.
+
+When a frame is being deleted, import its perspectives to another
+live frame if it exists.  Otherwise, if the frame being deleted
+is the last live frame, and emacs is running as a daemon, then
+store the perspectives of the frame being deleted so that they
+can be restored when a new frame is later created.
+"
+  (let ((min-num-frames (if (daemonp) 2 1)))
+    ;; Check whether the frame being deleted is the last visible frame.  In
+    ;; daemon mode, a fake fame is used, so the last visible frame occurs when
+    ;; the length of the frame list is 2.
+    (if (= (length (frame-list)) min-num-frames)
+        (when (daemonp)
+          (persp-save)
+          (setq persp-saved-hash perspectives-hash))
+      (persp-import-all-from-frame (get-other-frame) frame))))
+
+(defun persp-import-all-from-frame (to-frame from-frame)
+  "Import all perspectives from FROM-FRAME to TO-FRAME."
+  (let ((from-hash (with-selected-frame from-frame perspectives-hash))
+        (to-hash (with-selected-frame to-frame perspectives-hash)))
+    (cl-loop for name being the hash-keys of from-hash
+             when (not (gethash name to-hash)) do
+             (puthash name (gethash name from-hash)
+                      to-hash)))
```

Comment:
Rather than just ditching the old perspective if there already exists one with the same name, I think we should merge them as best we can.